packages/core/src/modules/oob/OutOfBandModule.ts

```diff
@@ -0,0 +1,297 @@
+      return this.connectionsModule.receiveInvitation(invitation)
+    }
+  }
+
+  /**
+   * Takes all messages from `requests~attach` attribute and pass them to the agent via event emitter.
+   *
+   * @param outOfBandMessage
+   */
+  public async receiveMessage(
+    outOfBandMessage: OutOfBandMessage,
+    config: ReceiveOutOfBandMessageConfig
+  ): Promise<ConnectionRecord | undefined> {
+    const { handshakeProtocols, services } = outOfBandMessage
+    const { autoAccept, reuse } = config
+    const messages = outOfBandMessage.getRequests()
+
+    if ((!handshakeProtocols || handshakeProtocols.length === 0) && (!messages || messages?.length === 0)) {
+      throw new AriesFrameworkError(
+        'One or both of handshake_protocols and requests~attach MUST be included in the message.'
+      )
+    }
+
+    if (handshakeProtocols) {
+      if (!this.areHandshakeProtocolsSupported(handshakeProtocols)) {
+        throw new AriesFrameworkError('Handshake protocols are not supported.')
+      }
+
+      let connectionRecord: ConnectionRecord
+      const existingConnection = await this.findExistingConnection(services)
+      if (existingConnection) {
+        this.logger.debug('Connection already exists.', { connectionId: existingConnection.id })
+        if (reuse) {
+          this.logger.debug('Reuse is enabled. Reusing an existing connection.')
+          connectionRecord = existingConnection
+          if (!messages) {
+            this.logger.debug('Out of band message does not contain any request messages.')
+            await this.sendReuse(outOfBandMessage, connectionRecord)
+          }
+        } else {
+          this.logger.debug('Reuse is disabled. Creating a new connection.')
+          connectionRecord = await this.createConnection(outOfBandMessage, autoAccept)
+        }
+      } else {
+        this.logger.debug('Connection does not exists. Creating a new connection.')
+        connectionRecord = await this.createConnection(outOfBandMessage, autoAccept)
+      }
+
+      if (messages) {
+        this.logger.debug('Out of band message contains request messages.')
+        if (!connectionRecord.isReady) {
+          // Wait until the connecion is ready and then pass the messages to the agent for further processing
+          this.connectionsModule
+            .returnWhenIsConnected(connectionRecord.id)
+            .then((c) => {
+              const connectionServices = c.theirDidDoc?.didCommServices
+              this.emitMessages(connectionServices, messages)
+            })
+            .catch((error) => {
+              if (error instanceof EmptyError) {
+                this.logger.warn(
+                  `Agent unsubscribed before connection got into ${ConnectionState.Complete} state`,
+                  error
+                )
+              } else {
+                this.logger.error('Promise waiting for the connection to be complete failed.', error)
+              }
+            })
```

Comment:
If the agent shuts down mid creating a connection, the messages are lost and won't be processed when starting the agent again. If we were to do this with a middleware of some kind (or listening to events) this wouldn't be dependant on the agent being alive during the whole process.

Maybe that is a fair tradeoff to make for now. If we create a connection, we could store the invitation in the connection record and resume later on. If we don't create a connection, we see it as more ephemeral and the agent must be online during the complete exchange.

Reply:
I think we should start storing the oob invitation in a didcomm message record (as we talked about during wg calls). This way we don't need a didexchange record and we keep a lot more flexibility going forward